File: Java/word-search.java
//  Word Search

// Time O(M^2*N^2), Space O(1)
class Solution {
    public boolean exist(char[][] board, String word) {
        if (board.length == 0)  return false;
        if (board[0].length == 0)   return false;
        for (int x = 0; x < board.length; x ++)
            for (int y = 0; y < board[0].length; y ++)
                if (dfs(board, word, x, y, 0))
                    return true;
        return false;
    }
    
    private boolean dfs(char[][] board, String word, int x, int y, int index) {
        if (index == word.length())   return true;
        if (x<0 || x>=board.length || y<0 || y>=board[0].length)    return false;
        if (board[x][y] != word.charAt(index))  return false;
        char temp = board[x][y];
        board[x][y] = '#';
        if (dfs(board, word, x-1, y, index+1)) {
            board[x][y] = temp;
            return true;
        }
        if (dfs(board, word, x+1, y, index+1)) {
            board[x][y] = temp;
            return true;
        }
        if (dfs(board, word, x, y-1, index+1)) {
            board[x][y] = temp;
            return true;
        }
        if (dfs(board, word, x, y+1, index+1)) {
            board[x][y] = temp;
            return true;
        }
        board[x][y] = temp;
        return false;
    }
}

File: leetcode/word-search.cc
//  Word Search

class Solution {
public:
    bool exist(vector<vector<char>>& board, string word) {
        if (board.size()==0 || board[0].size()==0)  
            return false;
        for (int i=0; i<board.size(); i++)
            for (int j=0; j<board[0].size(); j++)
                if (dfs(board, word, 0, i, j))
                    return true;
        return false;
    }
    
    bool dfs(vector<vector<char>> &board, string &word, int index, int x, int y) {
        if (index == word.size())   return true;
        if (x<0 || x>=board.size() || y<0 || y>=board[0].size())    
            return false;
        if (board[x][y] != word[index]) return false;
        char temp = board[x][y];
        board[x][y] = '*';
        if (dfs(board, word, index+1, x-1, y)) {
            board[x][y] = temp;
            return true;
        }
        if (dfs(board, word, index+1, x+1, y)) {
            board[x][y] = temp;
            return true;
        }
        if (dfs(board, word, index+1, x, y-1)) {
            board[x][y] = temp;
            return true;
        }
        if (dfs(board, word, index+1, x, y+1)) {
            board[x][y] = temp;
            return true;
        }
        board[x][y] = temp;
        return false;
    }
};
